### ISSUE TITLE
Address "The beatmap you're trying to submit isn't yours!" when creating a new map based on an existing mapset

### ISSUE THREAD

**Issue by iptq** [ppy/osu-stable-issues#341]
Continued from [this Twitter thread](https://twitter.com/Frontierrrrr/status/1220031401967308800).

For mappers trying to create new maps using audio and timing points from existing maps, the easiest thing to do is open the beatmap in the editor and hit File > Create new difficulty. This lets them create a map from scratch without extra effort. When they later try to create a mapset, they will invariably encounter the "The beatmap you're trying to submit isn't yours!" bug, unless they apply a well-known workaround.

The workaround comes in many forms but usually either involves some way of invalidating osu's knowledge of the old map and resetting BeatmapID and BeatmapsetID or duplicating the audio file elsewhere and dragging it back into osu and copying over the timing points. This requires messing with .osu files in ways that cause inconsistencies between the files and osu's internal map database.

Reproduction steps:

1. Open an already uploaded beatmap.
2. Create a new difficulty through the File menu.
3. Change the creator name through the song setup (F4) menu.
4. Delete the original difficulties from the main menu.
5. Try uploading the mapset from the newly created difficulty.

**Suggestion:** Could you add a button to the File menu of the editor, similar to the "Create new difficulty" button that creates a new mapset entirely? The button could bring up the same song setup (F4) dialog, ask the user for some details and then copy (or hardlink) the audio file and uninherited timing points to an entirely new mapset, and then updating osu's internal database to recognize the new mapset. Since this creates a fresh map and set, it should avoid the problem that the "The beatmap isn't yours!" dialog was meant to avoid, which I assume is people just changing the creator name and reuploading beatmaps. It will also save mappers from having to edit any .osu files in ways that osu isn't able to track.

I'm not sure how complex an addition like this could take, but almost all new mappers have run into this bizarre bug (until they learn of the workaround), so I'd appreciate if you could take some time and look into making this process easier.

---
- Was originally going to reopen #38 but it seems to be a bug specifically related to audio files containing numbers that was fixed already, just linking it here anyway.

**Comment by WinterIsPretty** [comment]
Why not just copy the song, load it in, and then insert the timing points into the osu file from the original file

**Comment by InvisibleSymbol** [comment]
> It will also save mappers from having to edit any .osu files in ways that osu isn't able to track.

Please read the issue before commenting

**Comment by WinterIsPretty** [comment]
I believe it disables you from adding a diff because you can copy a diff when you create a new one and that makes it so you can copy another person's beatmap

**Comment by nl-tatatat** [comment]
Yeah this would be a nice feature. I just copy the audio file from the song's folder and manually copy over all metadata and timing. Thats a lot of hassle though, and a lot to ask of new mappers to guess as well.

**Comment by iptq** [comment]
seeing how this is still a problem people are running into in the #mapping channel (and [this forum post](https://osu.ppy.sh/community/forums/topics/325639) was the second-most referenced url in the #modhelp channel), maybe a hotfix could just be to do some cache invalidation on wherever the beatmap id and set are being stored in the database, or only checking that the creator name matches the current username on upload?

**Comment by peppy** [comment]
Where is this flow being advertised as the correct way to use the editor?

This is intended behaviour. Creating a new map should be done by dragging an mp3 into the cleaner.

**Comment by iptq** [comment]
It's not being advertised, but there is also no obvious method for creating a new mapset while preserving timing (inherited), metadata, sometimes the background, and possibly hitsound samples. This is a workflow many mappers use to piggyback off an existing set.

The logical next step is to create a new difficulty, and then delete all of the existing maps which leads to this error. It's also possible to copy the timings and metadata manually, but it's a lot slower.

**Comment by peppy** [comment]
I guess the simple solution here is to forcefully clear all IDs from the `.osu` file when the creator string is changed from the song setup dialog. Does that sound like it would resolve the user flow you are seeing happen?

**Comment by iptq** [comment]
uhhh is there somewhere else the beatmap ID and beatmapset ID are stored? I've tried resetting the IDs through changing the .osu file directly when the client is closed but if the file remains in the same place, it would reset the next time I save the file / try to upload.

If you need an example of a workflow to test, it would look something like (1) create new difficulty, (2) change creator name through the song setup panel, (3) delete other difficulties, (4) submit. If this works, it should cover the majority of what new mappers are trying to do and anyone running into this problem otherwise is just doing something weird

**Comment by peppy** [comment]
The IDs would be restored automatically on startup if the hashes of the files have no changed.

**Comment by iptq** [comment]
Bump, people are still having this problem on the daily and newer mappers are still being taught to perform [this hack][1].

> forcefully clear all IDs from the .osu file when the creator string is changed from the song setup dialog

This sounds like it'll clear only the .osu file which is what players are already doing. I think the problem has to do with other places the IDs may be cached, like maybe the [osu!.db file][2].

> IDs would be restored automatically on startup

Where are these IDs being restored from?

[1]: https://osu.ppy.sh/community/forums/topics/325639
[2]: https://osu.ppy.sh/wiki/en/osu%21_File_Formats/Db_%28file_format%29#beatmap-information

**Comment by peppy** [comment]
To confirm, this is still happening in cases where the user is copying `.osu` files from another beatmap to their own by external means, correct?

**Comment by iptq** [comment]
That may happen, but I think the biggest use-case is:

> - Open an already uploaded
> - Create a new difficulty through the File menu.
> - Change the creator name through the song setup (F4) menu.
> - Delete the original difficulties from the main menu.
> - Try uploading the mapset from the newly created difficulty.

so essentially "turning" an existing uploaded mapset into a new one they want to upload.

I'd also like to note that tournament organizers would like to reupload existing maps with tweaks for their particular tournament needs (eg. AR8 edit, CS5 edit, removing a particular difficulty spike, etc.) that the mapper would not like to apply to their original map, but I'm sure most of them are familiar with the [workaround][1] and are using it.

[1]: https://osu.ppy.sh/community/forums/topics/325639

**Comment by BombBusiness** [comment]
> Reproduction steps:
> 
> 1. Open an already uploaded beatmap.
> 2. Create a new difficulty through the File menu.
> 3. Change the creator name through the song setup (F4) menu.
> 4. Delete the original difficulties from the main menu.
> 5. Try uploading the mapset from the newly created difficulty.

You don't need to delete the original difficulties. Instead, drag and drop the audio.mp3 file from the already uploaded beatmap. Copy and paste the metadata from your original difficulty into the new beatmapset. Finally, move the video (MV), thumbnail, and any other files from the old beatmap folder (excluding other beatmaps, of course) into the new beatmap folder.

Congrats, you can now upload your beatmap without anything blocking you.
